pyomo/contrib/gdpopt/nlp_solve.py

```diff
@@ -311,19 +312,17 @@ def detect_unfixed_discrete_vars(model):
 
 def preprocess_subproblem(m, config):
     """Applies preprocessing transformations to the model."""
-    # fbbt(m, integer_tol=config.integer_tolerance)
+    # First do FBBT
+    fbbt(m, integer_tol=config.integer_tolerance,
+         feasibility_tol=config.constraint_tolerance)
```

Comment:
FBBT may have some undesirable side effects. Tightened variable bounds can hurt NLP convergence in some cases. I recommend storing the variable bounds before calling `fbbt` and restoring the bounds at the end of `preprocess_subproblem` for any variables that do not get fixed. There should at least be an option to control that behavior, I think.

Reply:
So restoring bounds in combination with setting `deactivate_satisfied_constraints` to `True` can have problems. For example:
```
m = ConcreteModel()
m.x = Var(bounds=(-2, 10))
m.c1 = Constraint(expr=m.x >= 8)
m.obj = Objective(expr=m.x)

fbbt(m, deactivate_satisfied_constraints=True)

# This becomes an issue when we restore bounds: we've actually removed m.c1 from the model, and are relying on just 
# the variable bounds to enforce the constraint
print(m.c1.active) # False
```
So I think we either need to wait until after fbbt and use the (less smart) `deactivate_trivial_constraints` transformation or we can't restore the bounds to the originals.

Reply:
Good catch! I would be in favor of using the `deactivate_trivial_constraints` transformation.

Reply:
OK, sounds good! I think that makes sense.

Reply:
After thinking about this a bit more, the variable bounds should be restored right after `fbbt` rather than at the end of `preprocess_subproblem`. In general the variable bounds may impact later transformations.

Reply:
Oh, I think you're right that they should be restored before `deactivate_trivial_constraints` because we risk the same issue as above otherwise. It's not actually a problem now because `deactivate_trivial_constraints` isn't as smart as fbbt: It only checks individual constraints where the body is constant. So, with that design, if it's trivial with tight bounds it will be trivial with looser bounds. But I agree that in principle it should be called after restoring the bounds. However, the `detect_fixed_vars` transformation fixes variables whose bounds have closed, so it needs the tightened bounds. And I think `remove_zero_terms` is agnostic to the bounds. I'll switch to restoring them after `detect_fixed_vars`.

Reply:
Sounds good.